### DISCUSSION TITLE
Splink Altair graph saving

### DISCUSSION THREAD

**Discussion by illeamb** [moj-analytical-services/splink#857] [General]
I have been using the SPLINK package recently.

I was wondering how I can save some of the graphs that are generated from Splink. So the object that I am trying to print are  :

linker.match_weights_chart()
linker.m_u_parameters_chart()

They appear to be Vegalite objects. Is there an easy way to save them as a png file? I wasn't sure whether there was any documentation on how to save the objects. If I have missed it, feel free to point me in the right direction.

**Edit and Solution from maintainers 2023-06-16**
This question is now out of date due to [this PR](https://github.com/moj-analytical-services/splink/pull/1315)

All charts produced Splink are now full Altair Charts `alt.Chart()`.  You need to have Altair version 5 for them to work.

This means all charts can now be [more easily saved](https://altair-viz.github.io/user_guide/saving_charts.html) using e.g. `chart.save("myfile.png")`.  Note you can use file extensions `json`, `png`, `svg`, `pdf`, or `html` and others depending on what type output you want

Since all charts are now proper Altair charts, users experiencing difficulties should be asked to look at the [Altair documentation](https://altair-viz.github.io/user_guide/display_frontends.html).

**Comment by illeamb** [comment]
So further to that I had a look through the splink  source code and managed to save the objects as html file with the sample code below:

from splink import charts
plot_obj=restrictive_linker.m_u_parameters_chart()
charts.save_offline_chart(plot_obj,filename="test.html",overwrite=True)

I can see the charts in html format on the browser. How do I convert it to png through python?

**Reply by RobinL** [reply]
Sorry - this is a little more trouble than it should be due to us using some of the newest vega lite features that haven't yet made it into Altair on pypi

I think you need to install (see also the 'additional requirements')
https://github.com/altair-viz/altair_saver/

And then 
```
import altair_saver
c = linker.match_weights_chart()
altair_saver.save(c.spec, "saved.png")
```
Alternatively you could use the CLI described here:
https://vega.github.io/vega/usage/#:~:text=Back%20to%20reference-,Command%20Line%20Utilities,-The%20vega%2Dcli

to render the `c.spec` dictionary  by calling out to the command line

**Reply by illeamb** [reply]
Yeah, so tried something similar earlier and now tried your code snippet. I get the following error on both. I am using altair_saver version '0.5.0'.

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [12], line 3
      1 import altair_saver
      2 c = linker.match_weights_chart()
----> 3 altair_saver.save(c.spec, "saved.png")

File /usr/local/lib/python3.9/site-packages/altair_saver/_core.py:169, in save(chart, fp, fmt, mode, embed_options, method, suppress_data_warning, **kwargs)
    166 if embed_options is None:
    167     embed_options = alt.renderers.options.get("embed_options", None)
--> 169 Saver = _select_saver(method, mode=mode, fmt=fmt, fp=fp)
    170 saver = Saver(spec, mode=mode, embed_options=embed_options, **kwargs)
    172 return saver.save(fp=fp, fmt=fmt)

File /usr/local/lib/python3.9/site-packages/altair_saver/_core.py:69, in _select_saver(method, mode, fmt, fp)
     67         if s.enabled() and fmt in s.valid_formats[mode]:
     68             return s
---> 69     raise ValueError(f"No enabled saver found that supports format={fmt!r}")
     70 else:
     71     raise ValueError(f"Unrecognized method: {method}")

ValueError: No enabled saver found that supports format='png'

**Reply by RobinL** [reply]
See the additional requirements here https://github.com/altair-viz/altair_saver/

**Comment by NathanHGayle** [comment]
@illeamb Thanks for this thread. Update for future readers: Upon opening the saved HTML file there is a PNG download option.

**Reply by RobinL** [reply]
That's right, thanks. There are also some other options here if you need to automate: https://github.com/moj-analytical-services/splink/discussions/926#discussioncomment-4335509
